odo/backends/sql_csv.py

```diff
@@ -229,6 +230,101 @@ def append_csv_to_sql_table(tbl, csv, bind=None, **kwargs):
 
     kwargs = merge(csv.dialect, kwargs)
     stmt = CopyFromCSV(tbl, csv, bind=bind, **kwargs)
-    with bind.begin() as conn:
-        conn.execute(stmt)
+
+    if dialect == 'postgresql':
+        with bind.begin() as c:
+            with csv.open() as f:
+                c.connection.cursor().copy_expert(literal_compile(stmt), f)
+    else:
+        with bind.begin() as conn:
+            conn.execute(stmt)
     return tbl
+
+
+class NanIsNotNullFormatter(CSVFormatter):
+    def _save_chunk(self, start_i, end_i):
+
+        data_index = self.data_index
+
+        # create the data for a chunk
+        slicer = slice(start_i, end_i)
+        for i in range(len(self.blocks)):
+            b = self.blocks[i]
+            na_rep = 'nan' if b.dtype.kind == 'f' else self.na_rep
+            d = b.to_native_types(slicer=slicer,
+                                  na_rep=na_rep,
+                                  float_format=self.float_format,
+                                  decimal=self.decimal,
+                                  date_format=self.date_format,
+                                  quoting=self.quoting)
+
+            for col_loc, col in zip(b.mgr_locs, d):
+                # self.data is a preallocated list
+                self.data[col_loc] = col
+
+        ix = data_index.to_native_types(slicer=slicer, na_rep=self.na_rep,
+                                        float_format=self.float_format,
+                                        decimal=self.decimal,
+                                        date_format=self.date_format,
+                                        quoting=self.quoting)
+
+        pd.lib.write_csv_rows(self.data,
+                              ix,
+                              self.nlevels,
+                              self.cols,
+                              self.writer)
+
+
+# The set of sql dialects which do not bounce dataframes to disk to append
+# to a sql table.
+DATAFRAME_TO_TABLE_IN_MEMORY_CSV = frozenset({
+    'postgresql',
+})
+
+# TODO: figure out how to dynamically detect this number, users may have
+# compiled their sqlite with a different max variable number but this is the
+# default.
+SQLITE_MAX_VARIABLE_NUMBER = 999
```

Comment:
If you have compiled a larger max variable number then this will just chunk more than needed, if it is smaller it wont work. I have not encountered anyone with a smaller max variable number in practice, but we could add this as an arg to append_dataframe_to_sql_table.